### PR TITLE
Fix little-endian problem with hostid

### PR DESCRIPTION
**Pull request by jasom**
Fix for https://bugs.gentoo.org/show_bug.cgi?id=532746

### CODE REVIEW

File: gen_initramfs.sh [hunk 1]
```diff
@@ -100,7 +100,13 @@ append_base_layout() {
 	date -u '+%Y%m%d-%H%M%S' > ${TEMP}/initramfs-base-temp/etc/build_date
 	echo "Genkernel $GK_V" > ${TEMP}/initramfs-base-temp/etc/build_id
 
-	printf "$(hostid | sed 's/\([0-9A-F]\{2\}\)/\\x\1/gI')" > ${TEMP}/initramfs-base-temp/etc/hostid
+	little_endian="$(echo I | tr -d [:space:] | od -to2 | head -n1 | awk '{print $2}' | cut -c6)"
```
Comment (vapier) [inline gen_initramfs.sh]: this is more complicated than it needs to be.  you can use `file` on one of the installed files (like `/bin/sh`) to see whether it's big or little endian.

File: gen_initramfs.sh [hunk 2]
```diff
@@ -100,7 +100,13 @@ append_base_layout() {
...
+	if test "$little_endian" = 1; then
```
Comment (vapier) [inline gen_initramfs.sh]: we don't use `test`.  use `[[` instead.

File: gen_initramfs.sh [hunk 3]
```diff
@@ -100,7 +100,13 @@ append_base_layout() {
...
+	    hostid="$(hostid |sed 's/\(..\)\(..\)\(..\)\(..\)/\4\3\2\1/')"
```
Comment (vapier) [inline gen_initramfs.sh]: space before the `|`, and use the -E option to sed so you don't have to escape every `(` and `)`.

**Comment by jasom** [comment]
All of those changes (except the added space) are pretty much the exact opposite of how I typically code for shell.  Sorry for the style differences.

Suggested changes have now been applied.

File: gen_initramfs.sh [hunk 4]
```diff
@@ -100,7 +100,12 @@ append_base_layout() {
 	date -u '+%Y%m%d-%H%M%S' > ${TEMP}/initramfs-base-temp/etc/build_date
 	echo "Genkernel $GK_V" > ${TEMP}/initramfs-base-temp/etc/build_id
 
-	printf "$(hostid | sed 's/\([0-9A-F]\{2\}\)/\\x\1/gI')" > ${TEMP}/initramfs-base-temp/etc/hostid
+	if file -L /bin/sh | grep -q 'MSB executable'; then
```
Comment (vapier) [inline gen_initramfs.sh]: wouldn't you want to probe `${TEMP}/initramfs-base-temp/bin/sh` instead ?

File: gen_initramfs.sh [hunk 5]
```diff
@@ -100,7 +100,12 @@ append_base_layout() {
...
+	    hostid="$(hostid)"
+	else
+	    hostid="$(hostid | sed -E 's/(..)(..)(..)(..)/\4\3\2\1/')"
+	fi
+	printf "$(echo "$hostid" | sed 's/\([0-9A-F]\{2\}\)/\\x\1/gI')" > ${TEMP}/initramfs-base-temp/etc/hostid
```
Comment (vapier) [inline gen_initramfs.sh]: style nit: use braces with "${hostid}"

**Comment by vapier** [comment]
don't worry about getting things incorrect the first time.  the codebase is kind of a mess, and everyone has to start somewhere :).

**Comment by jasom** [comment]
More changes made.

**Comment by vapier** [comment]
thanks ... i fixed a few more style issues (bad use of spacing), and add missing `local hostid`, and added some comments explaining the issue.  see commit f3299dffb18d26fe500bc59a59a8bc591b2dcf78.

**Pull request closed without merging**